crates/matrix-sdk-ui/src/sync_service.rs

```diff
@@ -67,48 +67,44 @@ pub enum State {
     Error,
 }
 
-pub struct SyncService {
-    /// Room list service used to synchronize the rooms state.
+struct SyncServiceInner {
     room_list_service: Arc<RoomListService>,
-
-    /// Encryption sync taking care of e2ee events.
     encryption_sync_service: Arc<EncryptionSyncService>,
-
-    /// What's the state of this sync service?
     state: SharedObservable<State>,
-
-    /// Use a mutex everytime to modify the `state` value, otherwise it would be
-    /// possible to have race conditions when starting or pausing the
-    /// service multiple times really quickly.
-    modifying_state: AsyncMutex<()>,
-
-    /// Task running the room list service.
-    room_list_task: Arc<Mutex<Option<JoinHandle<()>>>>,
-
-    /// Task running the encryption sync.
-    encryption_sync_task: Arc<Mutex<Option<JoinHandle<()>>>>,
-
-    /// Global lock to allow using at most one `EncryptionSyncService` at all
-    /// times.
-    ///
-    /// This ensures that there's only one ever existing in the application's
-    /// lifetime (under the assumption that there is at most one
-    /// `SyncService` per application).
     encryption_sync_permit: Arc<AsyncMutex<EncryptionSyncPermit>>,
-
     /// Scheduler task ensuring proper termination.
     ///
     /// This task is waiting for a `TerminationReport` from any of the other two
     /// tasks, or from a user request via [`Self::stop()`]. It makes sure
     /// that the two services are properly shut up and just interrupted.
     ///
     /// This is set at the same time as the other two tasks.
-    scheduler_task: Arc<Mutex<Option<JoinHandle<()>>>>,
-
+    scheduler_task: Option<JoinHandle<()>>,
     /// `TerminationReport` sender for the [`Self::stop()`] function.
     ///
     /// This is set at the same time as all the tasks in [`Self::start()`].
-    scheduler_sender: Mutex<Option<Sender<TerminationReport>>>,
+    scheduler_sender: Option<Sender<TerminationReport>>,
+}
+
+pub struct SyncService {
+    inner: Arc<AsyncMutex<SyncServiceInner>>,
+
+    /// Room list service used to synchronize the rooms state.
+    room_list_service: Arc<RoomListService>,
+
+    /// What's the state of this sync service? This field is replicated from the
+    /// [`SyncServiceInner`] struct, but it should not be modified in this
+    /// struct. It's re-exposed here so we can subscribe to the state
+    /// without taking the lock on the `inner` field.
+    state: SharedObservable<State>,
+
+    /// Global lock to allow using at most one [`EncryptionSyncService`] at all
+    /// times.
+    ///
+    /// This ensures that there's only one ever existing in the application's
+    /// lifetime (under the assumption that there is at most one [`SyncService`]
+    /// per application).
+    encryption_sync_permit: Arc<AsyncMutex<EncryptionSyncPermit>>,
```

Comment:
Ditto, I don't see the point of having this cloned in both the outer and inner structs.

Reply:
d05784beb09098de4c5fd65d15c0f3dc91c89fe6